### PR TITLE
[declare] [rewrite] Use high-level declare API

### PR DESCRIPTION
**Pull request by ejgallego**

### CODE REVIEW

**Comment by ejgallego** [comment]
Maybe this should be included in #11818 ?

**Review by SkySkimmer — APPROVED** [review]
OK for plugin tutorial

**Comment by ejgallego** [comment]
CI is OK, PR is relatively straighforward IMHO; anyone would like to assign?

File: plugins/ltac/rewrite.ml [hunk 1]
```diff
@@ -1927,10 +1924,7 @@ let build_morphism_signature env sigma m =
   in
   let morph = e_app_poly env evd (PropGlobal.proper_type env) [| t; sig_; m |] in
   let evd = solve_constraints env !evd in
-  let evd = Evd.minimize_universes evd in
-  let m = Evarutil.nf_evars_universes evd (EConstr.Unsafe.to_constr morph) in
-  Pretyping.check_evars env evd (EConstr.of_constr m);
-  Evd.evar_universe_context evd, m
+  evd, morph
```
Comment (herbelin) [inline plugins/ltac/rewrite.ml:1927]: Is it obvious why the `check_evars` can now be skipped?

**Review by herbelin — COMMENTED** [review]
Is it pure reorganization or does it come with some change of semantics (e.g. around `check_evars`)?

**Comment by ejgallego** [comment]
> Is it pure reorganization or does it come with some change of semantics (e.g. around `check_evars`)?

The intent is to use the common code in `DeclareDef`, semantics should be the same but they can differ in some details sometimes; I think this case is OK.

`check_evars` is now performed in the call to `DeclareDef.prepare_definition` , together with the rest of the removed `minimize` etc...

File: plugins/ltac/rewrite.ml (hunk 1 above)
Comment (ejgallego) [reply plugins/ltac/rewrite.ml:1927]: Actually there is a minor difference, `DeclareDef` calls `check_evars_are_solved` whereas this code calls check_evars; I'm not sure what's going on here; see also #11984 .

IMO we should use a single call for all paths, unless a special check is needed, then I'd suggest we turn it into a parameter. (cc: @SkySkimmer )

Comment (ejgallego) [reply plugins/ltac/rewrite.ml:1927]: Other than that, I think semantics are the same.

File: plugins/ltac/rewrite.ml [hunk 2]
```diff
@@ -1892,14 +1892,11 @@ let declare_projection n instance_id r =
...
-  let term = EConstr.to_constr sigma term in
-  let cst = Declare.definition_entry ~types:typ ~univs term in
-  let _ : Constant.t =
-    Declare.declare_constant ~name:n ~kind:Decls.(IsDefinition Definition)
-      (Declare.DefinitionEntry cst)
+  let types = Some (it_mkProd_or_LetIn typ ctx) in
+  let kind, opaque, scope = Decls.(IsDefinition Definition), false, DeclareDef.Global Declare.ImportDefaultBehavior in
+  let impargs, udecl = [], UState.default_univ_decl in
+  let _r : GlobRef.t =
+    DeclareDef.declare_definition ~name ~scope ~kind ~opaque ~impargs ~udecl ~poly ~types ~body sigma
```
Comment (ejgallego) [inline plugins/ltac/rewrite.ml:1899]: Here there is another semantic change, the code now is doing universe restriction, etc... usually this is what we want; but if not we could add the proper mode to `declare_definition`

**Comment by ejgallego** [comment]
> CI is OK, PR is relatively straighforward IMHO; anyone would like to assign?

Any takers? This got stuck but IMO would be great go ahead with.

**Comment by Zimmi48** [comment]
@herbelin When leaving a partial review on a PR, make sure to either self-assign or re-request a review request from the relevant maintainers team. Otherwise, the PR won't appear anymore in the review request list of anyone and no one will notice that it needs an assignee.

File: doc/plugin_tutorial/tuto1/src/simple_declare.ml [hunk 3]
```diff
@@ -1,8 +1,6 @@
...
-  DeclareDef.declare_definition ~name ~scope ~kind ~impargs ?hook
-    ~opaque ~poly ~udecl ~types:tyopt ~body sigma
-
 let declare_definition ~poly name sigma body =
   let udecl = UState.default_univ_decl in
-  edeclare ~name ~poly ~scope:(DeclareDef.Global Declare.ImportDefaultBehavior)
-    ~kind:Decls.(IsDefinition Definition) ~opaque:false ~impargs:[] ~udecl sigma body None
+  let scope = DeclareDef.Global Declare.ImportDefaultBehavior in
+  let kind = Decls.(IsDefinition Definition) in
+  DeclareDef.declare_definition ~name ~scope ~kind ~impargs:[] ~udecl
```
Comment (ppedrot) [inline doc/plugin_tutorial/tuto1/src/simple_declare.ml:5]: Not directly related to this PR, but this kind of use shows that `impargs` should probably be an optional parameter.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
I am assigning. Modulo the potential subtle changes in semantics, this PR seems ready to merge, and I'll do so soon. @herbelin if you want more time to review please tell me.

File: doc/plugin_tutorial/tuto1/src/simple_declare.ml (hunk 3 above)
Comment (herbelin) [reply doc/plugin_tutorial/tuto1/src/simple_declare.ml:5]: About the status of impargs, I have #11272 which makes significant simplifications of the implicit arguments paths. It only depends on #11099 and #11135 now, so, its turn is soon coming.

**Comment by herbelin** [comment]
> When leaving a partial review on a PR, make sure to either self-assign or re-request a review request from the relevant maintainers team. Otherwise, the PR won't appear anymore in the review request list of anyone and no one will notice that it needs an assignee.

Ah, I did not know that. Sorry.

**Comment by herbelin** [comment]
> I am assigning. Modulo the potential subtle changes in semantics, this PR seems ready to merge, and I'll do so soon. @herbelin if you want more time to review please tell me.

The PR is ok to me as it is and I'm glad to see the `declare` path progressively unified and better understood. (About check_evars, I believe that some double-check will have to be done at some time, especially in relation with #11984, but that does not have to be in this PR.)

File: doc/plugin_tutorial/tuto1/src/simple_declare.ml (hunk 3 above)
Comment (ejgallego) [reply doc/plugin_tutorial/tuto1/src/simple_declare.ml:5]: It is quite hard to know when to turn a parameters into implicits; I understand large OCaml codebases actually consider implicits as evil as experimental evidence shows people just forget to use them when they are needed. I subscribe to this view.

In the particular case of `Declare` my plan is to use an `Info.t` record in the way it is done in Lemmas ; that IMHO provides the best of both approaches.

**Comment by ppedrot** [comment]
Let's merge then.

**Comment by ejgallego** [comment]
Merci!

**Pull request merged**